openmls/src/group/tests/test_proposal_validation.rs

```diff
@@ -656,92 +677,183 @@ fn test_valsem103(ciphersuite: Ciphersuite, backend: &impl OpenMlsCryptoProvider
-    } = validation_test_setup(PURE_PLAINTEXT_WIRE_FORMAT_POLICY, ciphersuite, backend);
-
-    // We now have alice create a commit. Then we artificially add an Add
-    // proposal with an existing identity (Bob).
-    let (_bob_credential_bundle, bob_key_package_bundle) =
-        generate_credential_bundle_and_key_package_bundle("Bob".into(), ciphersuite, backend);
-    let bob_key_package = bob_key_package_bundle.key_package().clone();
-
-    // Create the Commit.
-    let serialized_update = alice_group
-        .self_update(backend, None)
-        .expect("Error creating self-update")
-        .tls_serialize_detached()
-        .expect("Could not serialize message.");
+    for alice_and_bob_share_identities in [
+        KeyUniqueness::NegativeSameKey,
+        KeyUniqueness::PositiveSameKeyWithRemove,
+    ] {
+        // We now test if ValSem103 is also performed when a client receives a
+        // commit. Before we can test reception of (invalid) proposals, we set up a
+        // new group with Alice and Bob.
+        let ProposalValidationTestSetup {
+            mut alice_group,
+            mut bob_group,
+        } = validation_test_setup(PURE_PLAINTEXT_WIRE_FORMAT_POLICY, ciphersuite, backend);
+
+        // We now have alice create a commit. Then we artificially add an Add
+        // proposal with an existing identity (Bob).
+        let (_bob_credential_bundle, bob_key_package_bundle) =
+            generate_credential_bundle_and_key_package_bundle("Bob".into(), ciphersuite, backend);
+        let bob_key_package = bob_key_package_bundle.key_package().clone();
 
-    let plaintext = MlsMessageIn::tls_deserialize(&mut serialized_update.as_slice())
-        .expect("Could not deserialize message.")
-        .into_plaintext()
-        .expect("Message was not a plaintext.");
+        // Create the Commit.
+        let serialized_update = alice_group
+            .self_update(backend, None)
+            .expect("Error creating self-update")
+            .tls_serialize_detached()
+            .expect("Could not serialize message.");
 
-    // Keep the original plaintext for positive test later.
-    let original_plaintext = plaintext.clone();
+        let plaintext = MlsMessageIn::tls_deserialize(&mut serialized_update.as_slice())
+            .expect("Could not deserialize message.")
+            .into_plaintext()
+            .expect("Message was not a plaintext.");
+        // Keep the original plaintext for positive test later.
+        let original_plaintext = plaintext.clone();
 
-    let add_proposal = Proposal::Add(AddProposal {
-        key_package: bob_key_package,
-    });
+        let proposals = match alice_and_bob_share_identities {
+            KeyUniqueness::NegativeSameKey => {
+                let add_proposal = ProposalOrRef::Proposal(Proposal::Add(AddProposal {
+                    key_package: bob_key_package,
+                }));
+                vec![add_proposal]
+            }
+            _ => {
```

Comment:
Be explicit. The goal of this enum is to break the code whenever a new variant is introduced so that the implementer carefully updates test code